doc/design/bucketpath.rst

```diff
@@ -0,0 +1,157 @@
+==============================
+Design Document Bucket Path(s)
+==============================
+
+Introduction
+============
+
+- Purpose of the document / Problem description
+- Definitions and Acronyms
+
+Overview
+========
+
+Design Goals
+============
+
+Path Design
+===========
+
+Future Considerations
+=====================
+
+Sudo Code Outline
+=================
+
+.. code-block:: python
+
+    from exasol.bucketfs import PathBuilder
+
+    Path = PathBuilder(credentials)
+
+    udf_path = Path("bfsl://some/local/path/file.tar.gz")
+    http_bucket_path = Path("bfs://127.0.0.1:8888/service/bucket/some/file.tar.gz")
+    https_bucket_path = Path("bfss://127.0.0.1:8888/service/bucket/some/file.tar.gz")
+    chroot_path = Path("bfss://127.0.0.1:8888/service/bucket/some/sub/subsub/file.tar.gz?chroot=/some/sub/")
+
+
+
+.. code-block:: python
+
+    from typing import Protocol
+
+    # interface / protocol or abc
+    # Note: should mimic pathlib.Path so it can easily understood and adopted. 
+    class Pathlike(Protocol):
+
+        @property
+        def name:
+            ...
+
+        @property
+        def suffix:
+            ...
+
+        @property
+        def root:
+            ...
+
+        def as_uri():
+            ...
+
+        def exists():
+            ...
+
+        def is_dir():
+            ...
+
+        def is_file():
+            ...
+
+        def read_text():
+            ...
+
+        def read_bytes():
+            ...
+
+        def write_text():
+            ...
+
+        def write_bytes():
```

Comment:
this we need to orient more on what we are doing in the bucket, aka download and upload, because it is not simple reads and writes and especially we need to be able to read and write in chunks, but we can't support partial changes.

Reply:
Agree. I thought about the open two, but I am not sure if we can implement a FileObj, but in the end it doesn't need to be a full and real FileObj. Maybe, we don't allow seek and things and only implement the read and write.

Reply:
I now also remember what the issue with the FileObj was. We need to use a generator or iterator in requests.put to feed the data in chucks. There we have a pull principle for the data. However, FileObj write implements a push principle. That means, you probably need to spawn thread where the requests.put needs to run in and connect it with a queue to the write.

Reply:
I agree`open` does like impose additional considerations, so I would put that up for ... "potential improvements later".

Reply:
so, I would suggest that we add additional to read_* and write_* the function we use in the bucket, they are already known. Later, we can think about open.